app/master/build_scheduler.py

```diff
@@ -60,19 +60,24 @@ def needs_more_slaves(self):
             return False
         return True
 
-    def allocate_slave(self, slave):
+    def allocate_slave(self, slave: Slave) -> bool:
         """
         Allocate a slave to this build. This tells the slave to execute setup commands for this build.
-
-        :type slave: Slave
+        :param slave: The slave to allocate
+        :return: Whether slave allocation was successful; this can fail if the slave is unresponsive
         """
         if not self._build_started:
             self._build_started = True
             self._build.mark_started()
-        self._slaves_allocated.append(slave)
-        slave.setup(self._build, executor_start_index=self._num_executors_allocated)
+
+        # Increment executors before triggering setup. This helps make sure the build won't take down
+        # every slave in the cluster if setup calls fail because of a problem with the build.
+        next_executor_index = self._num_executors_allocated
         self._num_executors_allocated += min(slave.num_executors, self._max_executors_per_slave)
         analytics.record_event(analytics.BUILD_SETUP_START, build_id=self._build.build_id(), slave_id=slave.id)
+        self._slaves_allocated.append(slave)
+
+        return slave.setup(self._build, executor_start_index=next_executor_index)
```

Comment:
Ideally yes (actually ideally we'd just attempt the setup call before adjusting those values, then return if it fails.) It's a bit weird right now but as I mentioned in the comment, if for some reason the call to start setup fails because of something weird with the build then we want to make sure that won't take down every slave in the cluster. `self._num_executors_allocated` is the only thing preventing that currently.

This should definitely be improved, although it's a pretty rare occurrence that I don't believe we've seen in the wild. It's also gross because if the call to start setup does fail on every slave assigned to the build then the build will just hang forever in the building state. I'd like to come back to this and improve it but I'm not too concerned about it being a common issue.